docs/src/main/asciidoc/picocli.adoc

```diff
@@ -0,0 +1,178 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
+////
+= Quarkus - Command Mode with Picocli
+:extension-status: experimental
+
+include::./attributes.adoc[]
+
+https://picocli.info/[Picocli] is an open source tool for creating rich command line applications.
+
+Quarkus provides support for using Picocli. This guide contains examples of `picocli` extension usage.
+
+include::./status-include.adoc[]
+
+IMPORTANT: If you are not familiar with the Quarkus Command Mode, consider reading the link:command-mode-reference[Command Mode Applications] first.
+
+== Configuration
+
+Once you have your Quarkus project configured you can add the `picocli` extension
+to your project by running the following command in your project base directory.
+
+[source,bash]
+----
+./mvnw quarkus:add-extension -Dextensions="picocli"
+----
+
+This will add the following to your pom.xml:
+
+[source,xml]
+----
+    <dependency>
+        <groupId>io.quarkus</groupId>
+        <artifactId>quarkus-picocli</artifactId>
+    </dependency>
+----
+
+== Simple Picocli Application with one Command
+
+Simple PicocliApplication with only one Command can be created as follows:
+
+[source,java]
+----
+package com.acme.picocli;
+
+import picocli.CommandLine;
+
+import javax.enterprise.context.Dependent;
+import javax.inject.Inject;
+
+@CommandLine.Command // <1>
+public class HelloCommand implements Runnable {
+
+    @CommandLine.Option(names = {"-n", "--name"}, description = "Who will we greet?", defaultValue = "World")
+    String name;
+
+    private final GreetingService greetingService;
+
+    @Inject // <2>
+    public HelloCommand(GreetingService greetingService) {
+        this.greetingService = greetingService;
+    }
+
+    @Override
+    public void run() {
+        greetingService.sayHello(name);
+    }
+}
+
+@Dependent
+class GreetingService {
+    void sayHello(String name) {
+        System.out.println("Hello " + name + "!");
+    }
+}
+----
+<1> If there is only one class annotated with `picocli.CommandLine.Command` it will be used as entry point to Picocli CommandLine.
+<2> All classes annotated with `picocli.CommandLine.Command` are register as beans in CDI.
+
+== Picocli Application with multiple Commands
+
+When multiple classes has `picocli.CommandLine.Command` annotation, then one of them need to be also annotated with `io.quarkus.picocli.runtime.annotations.TopCommand`.
```

Comment:
having `@TopCommand` does not feel right - could it not reuse @QuarkusMain logic or at least similar logic to pick the main one ?

Reply:
maybe just use @Named ?

Reply:
I also have mixed feelings about `@TopCommand` (but perhaps I just need to get used to it...)
My preference would be that applications would always get access to the `CommandLine` instance, so they can configure it as needed and explicitly call `execute` on it (or `parseArgs` if they want to).
My approach was to skip `PicocliCommandLineFactory` altogether and let application code look like this:

```java
@QuarkusMain
@ApplicationScoped
@Command(name = "demo", mixinStandardHelpOptions = true)
public class ExampleApp implements Runnable, QuarkusApplication {
    @Inject
    PicocliBeansFactory factory;

    @Override
    public void run() {
        // business logic
    }

    @Override
    public int run(String... args) throws Exception {
        System.setProperty("quarkus.banner.enabled", "false");
        return new CommandLine(this, factory).execute(args);
    }
}
```

Reply:
@remkop Your example will work with current implementation. If user provide own `QuarkusMain` annotated class then `picocli` extension will not add `PicocliRunner` bean. I will add this to documentation.

I wanted simplest Quarkus/Picocli application to look like basic REST example. One, `Command` annotated, class could be good start for simple command line application.

As for `TopCommand`:
- I don't think that I can reuse `QuarkusMain` here, without big changes in `MainClassBuildStep`. `Command` class may just be `Object` and don't need to implement `QuarkusApplication`
- @maxandersen Using `Named` doesn't fell right also, since any `Object` can be annotated with `Command` and `Command` itself is not `Qualifier`.
- We could add `name` to `TopCommand` and have similar logic as in `QuarkusMain`. In fact there is already `quarkus.picocli.top-command` property which will work in similar way. It takes FQCN of class which will be used as entry point. Not sure if this property should be runtime or build time configuration. For now it is runtime.
- We could drop `TopCommand` all together and add `PicocliRunner` if there is only one `Command` annotated class. In all other cases user would be forced to provided `QuarkusMain` on its own.
- Or just drop `PicocliRunner` entirely as @remkop suggested.

Reply:
Personally I like the annotation based approach. I think you can make the @QuarkusMain integration work without major MainClassBuildStep changes by looking for classes that have both @QuarkusMain and @Command. If one is found then make PicoliRunner use that class, and create a QuarkusApplicationClassBuildItem so that Quarkus won't look for the QuarkusMain class itself.

Reply:
@stuartwdouglas, if I understand correctly, you propose that instead of introducing a new `@TopCommand` annotation, applications would use both the combination of the `@QuarkusMain` and the `@Command` annotation to specify a main class. That makes sense to me.

Reply:
Yea. 

The only real downside to this is that it could create confusion, as it is a 3rd possible way to use @QuarkusMain, but it is still doing pretty much the same thing so I think it should be ok.

Reply:
@stuartwdouglas So picocli extension would need to have copy of logic for finding `@QuarkusMain` and if this Main is also annotated with `Command` then it would produce QuarkusApplicationClassBuildItem pointing to `PicocliRunner`. But, I think that with such approach `PicocliRunner` would need to be generated class. Both `@QuarkusMain` and `@Command` are not Qualifiers, so I'm not sure how to pass it to current `PicocliRunner`. This would also mean that QuarkusApplicationClassBuildItem has priority. If it exist then MainClassBuildStep will always use it. 

Note that current approach, with `@TopCommand` qualifiers, also allows things like this (after I add `METHOD` `@Target`): 

```java
@ApplicationScoped
public class Config {

    @Produces
    @TopCommand
    @IfBuildProfile("dev")
    public Object devCommand() {
        // Since this is Object, it can be also Class<?>
        // Instance will be created by CommandLine.IFactory
        return DevCommand.class;
    }

    @Produces
    @TopCommand
    @IfBuildProfile("prod")
    public Object prodCommand() {
        return ProdCommand.class;
    }

}
````

Reply:
That is a good point that this allows for producer methods. It also allows for the 'Right click and run from IDE' use case, if you do Quarkus.run(PicocliRunner.class) in your main method, which my approach would not allow (we should probably document this). 

Thinking about it I think that the current approach is probably best.